### ISSUE TITLE
Support fixed bucket sizes

### ISSUE THREAD

**Issue by va3093** [typesense/typesense#1943]
## TL;DR
Support for chunking responses into buckets with flexible sizes, and reordering the results within each bucket based on secondary sorting criteria.

## Description

Consider the following query:

```bash
curl -X POST "https://xxx.a1.typesense.net/multi_search" \
-H "X-TYPESENSE-API-KEY: xxx" \
-H "Content-Type: application/json" \
-d '{
  "searches": [
    {
      "collection": "content",
      "q": "us election",
      "query_by": "metadata.name",
      "sort_by": "_text_match(buckets:10):desc,metadata.stars.total:desc"
    }
  ]
}'
```

Depending on the data in the database, this could return either 1000 results or 20 results. In either case, we want to prioritize content with higher popularity on our site. However, it's challenging to pick a fixed number of buckets that will provide the desired behavior in both extremes. For example, if we choose 20 buckets, an item ranked 50th by text_match in the 1000 results case might be boosted to the top, while an item in a 20-result case might receive no boost at all.

### Static Bucket Sizes

Clients can define a fixed bucket size. If we set a bucket size of 20, we should achieve adequate boosting for results with similar text_match scores in the 1000-results case. In the 20-results case, users would still see the most popular content first.

### Dynamic Bucket Sizes Based on text_match Variance

An ideal solution would involve dynamically grouping results based on their `text_match` score variance. This approach adapts to different types of result sets, delivering a more meaningful ranking by either re-sorting results with similar scores or leaving them as-is when the variance is high.

#### Example:

Imagine a search query for "us election" that returns the following results, each with a `text_match` score and a `popularity` metric (such as the number of stars or views):

- **Result A**: `text_match: 98`, `popularity: 10`
- **Result B**: `text_match: 97`, `popularity: 50`
- **Result C**: `text_match: 95`, `popularity: 40`
- **Result D**: `text_match: 80`, `popularity: 30`
- **Result E**: `text_match: 75`, `popularity: 5`

In this case, results A, B, and C have similar `text_match` scores (highly relevant to the search). Since their relevance is so close, we can apply secondary sorting by `popularity`, giving us this order:

1. **Result B**: `text_match: 97`, `popularity: 50`
2. **Result C**: `text_match: 95`, `popularity: 40`
3. **Result A**: `text_match: 98`, `popularity: 10`

Results D and E, however, have much lower `text_match` scores compared to the top three results. Because there is a large variance in `text_match`, re-sorting by `popularity` would not make sense for these items. Therefore, these results remain in their original positions, even though Result E has lower popularity:

4. **Result D**: `text_match: 80`, `popularity: 30`
5. **Result E**: `text_match: 75`, `popularity: 5`

#### Explanation:

- **Low text_match variance**: When the `text_match` scores are close (e.g., Results A, B, and C), the secondary sorting by `popularity` makes sense to further differentiate the results.
- **High text_match variance**: When there's a large gap in `text_match` scores (e.g., Results D and E compared to the top three), re-sorting by `popularity` is not applied, as these results are already less relevant.

By adjusting bucket sizes dynamically based on the variance in `text_match` scores, this method ensures that results are grouped and re-sorted only when it makes sense, leading to a more relevant and meaningful ranking for the user.

**Comment by ajgeiss0702** [comment]
I was surprised that, at the very least, a fixed bucket size was not already in type sense, as I cannot think of a use case where a fixed number of buckets is useful due to the possibility of the large differences in the number of search results.

I hope that one of these solutions can be implemented to help improve search ordering

**Comment by Clasyc** [comment]
Yes, this would be very useful for our project. At the moment, we don't have an option to implement proper popularity ratings.

**Comment by dziugas-liaudinskas** [comment]
This blocks us too.

**Comment by Clasyc** [comment]
@kishorenc apologies for the ping. What do you guys think about this feature? Is it under consideration, or has it been postponed to future releases?

**Comment by jasonbosco** [comment]
This is now available in `28.0.rc36`:

```
"sort_by": "_text_match(buckets_size:10):desc,metadata.stars.total:desc"
```
